File: ms-chats/src/models/Message.ts
import {
  Entity,
  Column,
  CreateDateColumn,
  UpdateDateColumn,
  PrimaryColumn,
  JoinColumn,
  ManyToOne,
} from "typeorm";

import { v4 as uuid } from "uuid";
import { Chat } from "./Chat";
import { User } from "./User";

@Entity("messages")
export class Message {
  @PrimaryColumn()
  id: string;

  @JoinColumn({ name: "chat_id" })
  @ManyToOne(() => Chat)
  chat: Chat;

  @JoinColumn({ name: "sender_id" })
  @ManyToOne(() => User)
  user: User;

  @Column()
  chat_id: string;

  @Column()
  sender_id: string;

  @Column()
  content: string;

  @CreateDateColumn()
  created_at: Date;

  @UpdateDateColumn()
  updated_at: Date;

  constructor() {
    if (!this.id) {
      this.id = uuid();
    }
  }
}

File: ms-chats/models/message.go
package models

import (
	"errors"
	"time"

	uuid "github.com/satori/go.uuid"
	"gorm.io/gorm"
)

type Message struct {
	ID           string `json:"id" gorm:"column:id;" sql:"type:uuid;primary_key;default:uuid_generate_v4()"`
	ChatId       string `json:"chat_id"`
	Chat         Chat   `gorm:"foreignKey:ChatId;references:id"`
	SenderId     string `json:"sender_id"`
	Content      string `json:"content"`
	CreatedAt    time.Time `json:"created_at"`
	UpdatedAt    time.Time `json:"updated_at"`
	DeletedAt    gorm.DeletedAt `gorm:"index" json:"deleted"`
}

func (m *Message) BeforeCreate(tx *gorm.DB) (err error) {
	m.ID = uuid.NewV4().String()

	if m.ID == "" {
	  err = errors.New("can't save invalid data")
	}
	return
}